pkg/cache/internal/informers_map.go

```diff
@@ -213,24 +242,44 @@ func (ip *specificInformersMap) Get(ctx context.Context, gvk schema.GroupVersion
 	return started, i, nil
 }
 
-func (ip *specificInformersMap) addInformerToMap(gvk schema.GroupVersionKind, obj runtime.Object) (*MapEntry, bool, error) {
+func (ip *InformersMap) informersByType(obj runtime.Object) map[schema.GroupVersionKind]*MapEntry {
+	switch obj.(type) {
+	case *unstructured.Unstructured, *unstructured.UnstructuredList:
+		return ip.informers.Unstructured
+	case *metav1.PartialObjectMetadata, *metav1.PartialObjectMetadataList:
+		return ip.informers.Metadata
+	default:
+		return ip.informers.Structured
+	}
+}
+
+func (ip *InformersMap) addInformerToMap(gvk schema.GroupVersionKind, obj runtime.Object) (*MapEntry, bool, error) {
 	ip.mu.Lock()
 	defer ip.mu.Unlock()
 
 	// Check the cache to see if we already have an Informer.  If we do, return the Informer.
 	// This is for the case where 2 routines tried to get the informer when it wasn't in the map
 	// so neither returned early, but the first one created it.
-	if i, ok := ip.informersByGVK[gvk]; ok {
+	if i, ok := ip.informersByType(obj)[gvk]; ok {
 		return i, ip.started, nil
 	}
 
 	// Create a NewSharedIndexInformer and add it to the map.
-	var lw *cache.ListWatch
-	lw, err := ip.createListWatcher(gvk, ip)
+	lw, err := ip.makeListWatcher(gvk, obj)
 	if err != nil {
 		return nil, false, err
 	}
-	ni := cache.NewSharedIndexInformer(lw, obj, resyncPeriod(ip.resync)(), cache.Indexers{
+	ni := cache.NewSharedIndexInformer(&cache.ListWatch{
+		ListFunc: func(opts metav1.ListOptions) (runtime.Object, error) {
+			ip.selectors(gvk).ApplyToList(&opts)
+			return lw.ListFunc(opts)
+		},
+		WatchFunc: func(opts metav1.ListOptions) (watch.Interface, error) {
+			ip.selectors(gvk).ApplyToList(&opts)
+			opts.Watch = true // Watch needs to be set to true separately
+			return lw.WatchFunc(opts)
+		},
```

Comment:
Q: would it make sense to do this in `makeListWatcher` so that `makeListWatcher` alredy creates the final `ListWatch` vs having "layered" ListWatches?

Reply:
Originally it was in each list watcher creation function, although the code was duplicate between each watch/listfunc, which was a bit repetitive. Do you see another way to do this that doesn't end up in repetition?

Reply:
I see the following options:
1. write the listFunc and watchFunc in makeListWatcher inside the switch in variables and then build the ListWatch like it is done here at the end of the makeListWatcher func
1. writing the ListWatch'es in makeListWatcher in a var and then add this code at the end of the makeListWatcher func
1. defining inline modifyListOptions modifyWatchOptions func in makeListWatcher and calling them in each case

I would probably go with option 1, but also fine to just keep it as is if you want to